### ISSUE TITLE
A port that doesn't rely on node-gyp?

### ISSUE THREAD

**Issue by ffMathy** [uhop/node-re2#18]
Node-gyp can be quite a pain to use on Windows.

Is it possible to make it node-gyp independent?

**Comment by uhop** [comment]
I am open to suggestions, and working pull requests.

**Comment by uhop** [comment]
It looks like this question is more properly addressed to Node developers. There is so-called [N-API](https://nodejs.org/api/n-api.html), which may be better suited for Windows. It is deemed "stable", but only in the latest versions of Node.

**Comment by uhop** [comment]
Now with Node 6 is out of LTS we are closer to implement `re2` with [N-API](https://nodejs.org/api/n-api.html). The doc still has quite a few sections marked "Experimental" (for example, "Life Cycle") yet we can start exploring.

**Comment by viceice** [comment]
What the state here now, since node v10 is out of lts?

**Comment by uhop** [comment]
Not started yet.

**Comment by Richienb** [comment]
We can just use Wasm. This way, it will always be supported in Node.js and browsers will support it too.

**Comment by uhop** [comment]
All good ideas. Just keep in mind:

* The existing code uses the only bindings available at the time of creation: C++ with Nan using node-gyp.
* N-API is a C bindings, which requires rewriting the binding code.
  * Let me stress: it is not a matter of "recompile and it just works".
* WASM is used to run programs compiled from other languages to a specific target.
  * `re2` is a component, a class, not a complete runnable program.
  * There are no bindings to interface JavaScript objects or impersonate them.

I was assured that the proper WASM bindings are coming, but so far I didn't see them even in the alpha stage. Given that the main driving force behind WASM were Mozilla guys, I worry that their efforts will be delayed significantly by the recent Mozilla reorg.

Required reading:

* [C++ addons](https://nodejs.org/api/addons.html)
* [N-API](https://nodejs.org/api/n-api.html)
* [WASI](https://nodejs.org/api/wasi.html)

But I love to proven wrong, and, as always, patches are welcome!

**Comment by uhop** [comment]
Another thing to throw in the pile of possible ports: [deno](https://deno.land/) support. It is supposed to provide a way to add extensions based on [Rust](https://www.rust-lang.org/) (another Mozilla technology) and it should support WebAssembly in some form. So far nothing was made public, but it is clearly coming at some point.

**Comment by uhop** [comment]
Yet another thought: the original premise ("Node-gyp can be quite a pain to use on Windows") is more-or-less obsolete now. Starting with 1.14 `re2` pre-builds binaries for multiple platforms including Mac and Windows. It is conceivable to use `re2` on Windows without installing Python and the C++ toolchain.

It makes the Deno support more interesting for a simple reason: it is a completely unsupported platform, while other ways to support Node are largely irrelevant for end users &mdash; after setting it up once there is no difference how `re2` was compiled.

**Comment by Richienb** [comment]
@uhop Rust WASM supports those things and C++ simultaneously.

**Comment by uhop** [comment]
What things exactly? Component bindings? I would love to see how.

**Comment by Richienb** [comment]
@uhop Most of the re2 binding logic has already been done in https://github.com/nickdesaulniers/rust-re2. All we need to do is add type declarations. https://rustwasm.github.io/docs/wasm-bindgen/contributing/design/exporting-rust-struct.html?highlight=exporting,struct#exporting-a-struct-to-js

**Comment by uhop** [comment]
It is nice to know that somebody took care of things for me. I will be cheering for that projects on sidelines.

**Comment by uhop** [comment]
It looks like alternative ways to build a binary extensions are more mature now. Specifically WASM looks like an interesting option.

Closing this issues for two reasons:

* There are alternative projects based on WASM.
* All chrome code that bridges RE2 with JS should be rewritten, which requires a new project.

**Comment by viceice** [comment]
looks like all other wasm projects are dead or outdated, so they're not a drop-in replacement for `RegExp` 😞 

maybe this can help to port?
- https://github.com/toyobayashi/emnapi

**Comment by uhop** [comment]
Granted that I checked it some time ago and the situation could change by now: one problem with `wasm` was that it doesn't support strings as data type. Sending a string as an array of numbers immediately negates any performance benefits. I don't know what the story is with strings-specific `wasm` commands now.

Ideally we shouldn't send strings at all, we should share them with `wasm` code. And there were such proposals on shared buffers. I don't know how far they were advanced.

The last time I checked `wasm` was not suitable for a performant string processing. I would be happy proven wrong. I would dance in joy if text processing is not an after-thought anymore, buffers and efficient string commands are there and their support is universal.
